Assignment 2 - YakimS

YakimS opened the pull request.

YakimS committed:
Upload Assignment 2 files

brzozan commented on YakimS/assignment2/BinaryTree.java:
```diff
@@ -0,0 +1,231 @@
+package CodeUAss1Package;
+
+import javax.management.openmbean.KeyAlreadyExistsException;
```
It is better not to depend on such complex packages just for exception. Better use something from `java.lang`, like `IllegalArgumentException`.

brzozan commented on YakimS/assignment2/BinaryTree.java:
```diff
@@ -0,0 +1,231 @@
+package CodeUAss1Package;
+
+import javax.management.openmbean.KeyAlreadyExistsException;
+import java.util.*;
```
Please avoid importing via asterisk.

brzozan commented on YakimS/assignment2/BinaryTree.java:
```diff
@@ -0,0 +1,231 @@
+public class BinaryTree<T> {
+
+    private Node root;
+    private Set<Integer> keySet;
```
The whole tree has generic parameter `<T>`, so `keySet` should probably also use it.

YakimS replied:
The generic parameter is referred to the type of data the tree stores in his nodes, and I choose to store the key as integers. Can you please explain why should I use the keys of the nodes to be the same type as the value they reference?

YakimS replied:
You've answered my question at the next change request.

brzozan commented on YakimS/assignment2/BinaryTree.java:
```diff
@@ -0,0 +1,231 @@
+        printBinaryTree(this.root,0);
+    }
+    private void printBinaryTree(Node root, int level){
+        if(root==null)
```
Please avoid `if`s, `for`s etc without curly braces. It is hard to read and error-prone.

brzozan commented on YakimS/assignment2/BinaryTree.java:
```diff
@@ -0,0 +1,231 @@
+     * @param key - key of the added node
+     * @throws KeyAlreadyExistsException if there is already a node with key as given in the parameter
+     */
+    public void add(T value, int key) throws KeyAlreadyExistsException
```
Key is not really needed, since it is not intended to be a map. Better to just use value as the key.

brzozan commented on YakimS/assignment2/BinaryTree.java:
```diff
@@ -0,0 +1,231 @@
+        *  to that null current node
+        */
+        Node currentNode = root;
+        Random random = new Random();
```
Using a randomized data structure is an interesting idea, but it makes it very hard to test and to reason about. Unit tests are not required for this assignment, but if they were, it would be nearly impossible to check if they are correct, and since you use `Random` without seed, the results you will get are very likely to be different each time you run the program.

brzozan commented on YakimS/assignment2/BinaryTree.java:
```diff
@@ -0,0 +1,231 @@
+    public int getLowestCommonAncestor(int key1, int key2) throws KeyNotFoundException {
+
+        // create paths to the two keys given
+        List<Integer> firstKeyPath = getKeyPath(key1, root, new LinkedList<>());
```
The assignment explicitly asked to avoid storing nodes in additional data structures. While technically you are storing keys instead of nodes, in practice it is the same. Could you change the code to avoid that?

brzozan reviewed (changes requested):
Thank you for the PR!

Top-level comment: please format code consistently - braces placed always in the same way, spaces used the same way between keywords and parentheses etc.
Currently it is quite difficult to read, because formatting varies between methods and statements.

I added a few more detailed comments in code.

YakimS committed:
Delete KeyNotFoundException.java

YakimS committed:
Update BinaryTree

YakimS committed:
Update Main.java

YakimS committed:
Uplode TreeNode

YakimS commented:
I have assigned the requested changes.
Thank you very much for the review!

brzozan commented on YakimS/assignment2/BinaryTree.java:
```diff
@@ -124,108 +71,83 @@ else if(printAncestors(key,root.getLeftSon()) ||
-            throw new KeyNotFoundException();
+            throw new IllegalArgumentException("Value not found");
+        }
+        else if(nodeInPathOfVal1.equals(nodeInPathOfVal2))
```
Optional: At this point you are operating only on nodes that are already in the tree and do not change, so it is OK to just use `==` for comparison.

brzozan commented on YakimS/assignment2/BinaryTree.java:
```diff
@@ -124,108 +71,83 @@ else if(printAncestors(key,root.getLeftSon()) ||
+            {
+                return nodeInPathOfVal1.getValue();
+            }
+            else if(nodeInPathOfVal1.getLevel() < nodeInPathOfVal2.getLevel())
```
Optional: No need for `else` here. Having two separate ifs (one for node equality, one for level calculations) will be a bit easier to read, since they are logically separate.

brzozan commented on YakimS/assignment2/BinaryTree.java:
```diff
@@ -124,108 +71,83 @@ else if(printAncestors(key,root.getLeftSon()) ||
             return null;
         }
-        if(root.getKey() == key)
+        else if(root.getValue() == value1)
```
This should be `.equals`.  Two `Integer` (or `String`, or ...) objects can carry the same value, but not be equal when compared via `==`.

brzozan reviewed (approved):
Thank you for applying requested changes!

I have a few optional comments and one non-optional (non-optional, because it is a bug).

I am approving this PR, but before merging please fix the non-optional comment. Thank you!

YakimS committed:
Upload Assignment2 files

Thank you very much Michal!
I've made the optional and non-optional changes to the code, and documented some of your remark so hopefully i wont repeat these misakes.
